File: ios/Classes/FlutterIzipayPlugin.swift
import Flutter
import IzipayPayButtonSDK
import MastercardSonic
import UIKit
import VisaSensoryBranding

public class FlutterIzipayPlugin: NSObject, FlutterPlugin, IzipayPaymentDelegate {
  private var eventSink: FlutterEventSink?

  private let sonicManager: MCSonicController = MCSonicController()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(
      name: "flutter_izipay/method_channel", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(
      name: "flutter_izipay/event_channel", binaryMessenger: registrar.messenger())

    let instance: FlutterIzipayPlugin = FlutterIzipayPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "openFormToSaveCard":
      if let rootViewController = UIApplication.shared.windows.first?.rootViewController {
        rootViewController.overrideUserInterfaceStyle = .light
      }

      guard let args = call.arguments as? [String: String] else { return }

      openFormToSaveCard(data: args)
    case "payWithCard":
      if let rootViewController = UIApplication.shared.windows.first?.rootViewController {
        rootViewController.overrideUserInterfaceStyle = .light
      }

      guard let args = call.arguments as? [String: String] else { return }

      payWithCard(data: args)
    case "payWithYape":
      if let rootViewController = UIApplication.shared.windows.first?.rootViewController {
        rootViewController.overrideUserInterfaceStyle = .light
      }

      guard let args = call.arguments as? [String: String] else { return }

      payWithYape(data: args)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  func openFormToSaveCard(data: [String: String]) {
    print("openFormToSaveCard: \(data)")
  }

  func payWithCard(data: [String: String]) {
    let date = Date()
    let milliseconds = Int(date.timeIntervalSince1970 * 1000)

    var configPayment = ConfigPaymentIzipay()
    configPayment.enviroment = data["environment"]
    configPayment.action = "pay"
    configPayment.publicKey = data["publicKey"]
    configPayment.transactionId = data["transactionId"]
    configPayment.merchantCode = data["merchantCode"]

    configPayment.order = OrderPaymentIzipay()
    configPayment.order?.orderNumber = data["orderNumber"]
    configPayment.order?.currency = "PEN"
    configPayment.order?.amount = data["amount"]
    configPayment.order?.payMethod = [PaymentMethodIzipay.card]

    configPayment.order?.processType = "autorize"
    configPayment.order?.merchantBuyerId = data["userId"]
    configPayment.order?.dateTimeTransaction = "\(milliseconds)"

    configPayment.token = TokenPaymentIzipay()

    configPayment.billing = BillingPaymentIzipay()
    configPayment.billing?.firstName = data["firstName"]
    configPayment.billing?.lastName = data["lastName"]
    configPayment.billing?.email = data["email"]
    configPayment.billing?.phoneNumber = data["phoneNumber"]
    configPayment.billing?.street = data["street"]
    configPayment.billing?.city = data["city"]
    configPayment.billing?.state = data["state"]
    configPayment.billing?.country = data["country"]
    configPayment.billing?.postalCode = data["postalCode"]
    configPayment.billing?.documentType = data["documentType"]
    configPayment.billing?.document = data["documentNumber"]

    configPayment.shipping = ShippingPaymentIzipay()

    configPayment.appearance = AppearencePaymentIzipay()
    configPayment.appearance?.language = "ESP"
    configPayment.appearance?.formControls = AppearenceControlsPaymentIzipay()
    configPayment.appearance?.formControls?.isAmountLabelVisible = true
    configPayment.appearance?.formControls?.isLangControlVisible = false
    configPayment.appearance?.visualSettings = AppearenceVisualSettingsPaymentIzipay()
    configPayment.appearance?.visualSettings?.presentationMode = .fullscreen
    configPayment.appearance?.theme = "purple"
    configPayment.appearance?.customTheme = CustomThemePaymentIzipay()
    configPayment.appearance?.logo = data["logoUrl"]

    let sensoryBrandingVisa = SensoryBranding()
    sensoryBrandingVisa.backdropColor = .white
    sensoryBrandingVisa.isSoundEnabled = true
    sensoryBrandingVisa.isHapticFeedbackEnabled = true
    sensoryBrandingVisa.checkmarkMode = .checkmark

    let sensoryBrandMastercard = MCSonicView()
    sensoryBrandMastercard.background = MCSonicBackground.white

    configPayment.appearance?.sensoryBrand = AppearenceSensoryBrandIzipay()
    configPayment.appearance?.sensoryBrand?.visaSBView = sensoryBrandingVisa
    configPayment.appearance?.sensoryBrand?.mastercardSBView = sensoryBrandMastercard

    configPayment.urlIPN = data["webhookUrl"]

    let s = UIStoryboard(name: "IziPayment", bundle: Bundle(for: PaymentFormViewController.self))
    let vc =
      s.instantiateViewController(withIdentifier: "PaymentFormView") as! PaymentFormViewController
    vc.delegate = self
    vc.configPayment = configPayment
    vc.modalPresentationStyle = .fullScreen

    guard
      let rootViewController = UIApplication.shared.connectedScenes
        .compactMap({ $0 as? UIWindowScene })
        .flatMap({ $0.windows })
        .first(where: { $0.isKeyWindow })?.rootViewController
    else {
      print("No se pudo obtener el rootViewController")
      return
    }

    DispatchQueue.main.async {
      rootViewController.present(vc, animated: true, completion: nil)
    }
  }

  func payWithYape(data: [String: String]) {
    
    let date = Date()
    let milliseconds = Int(date.timeIntervalSince1970 * 1000)

    var configPayment = ConfigPaymentIzipay()
    configPayment.enviroment = data["environment"]
    configPayment.action = "pay"
    configPayment.publicKey = data["publicKey"]
    configPayment.transactionId = data["transactionId"]
    configPayment.merchantCode = data["merchantCode"]

    configPayment.order = OrderPaymentIzipay()
    configPayment.order?.orderNumber = data["orderNumber"]
    configPayment.order?.currency = "PEN"
    configPayment.order?.amount = data["amount"]
    configPayment.order?.payMethod = [PaymentMethodIzipay.yape_code]

    configPayment.order?.processType = "autorize"
    configPayment.order?.merchantBuyerId = data["userId"]
    configPayment.order?.dateTimeTransaction = "\(milliseconds)"

    configPayment.token = TokenPaymentIzipay()

    configPayment.billing = BillingPaymentIzipay()
    configPayment.billing?.firstName = data["firstName"]
    configPayment.billing?.lastName = data["lastName"]
    configPayment.billing?.email = data["email"]
    configPayment.billing?.phoneNumber = data["phoneNumber"]
    configPayment.billing?.street = data["street"]
    configPayment.billing?.city = data["city"]
    configPayment.billing?.state = data["state"]
    configPayment.billing?.country = data["country"]
    configPayment.billing?.postalCode = data["postalCode"]
    configPayment.billing?.documentType = data["documentType"]
    configPayment.billing?.document = data["documentNumber"]

    configPayment.shipping = ShippingPaymentIzipay()

    configPayment.appearance = AppearencePaymentIzipay()
    configPayment.appearance?.language = "ESP"
    configPayment.appearance?.formControls = AppearenceControlsPaymentIzipay()
    configPayment.appearance?.formControls?.isAmountLabelVisible = true
    configPayment.appearance?.formControls?.isLangControlVisible = false
    configPayment.appearance?.visualSettings = AppearenceVisualSettingsPaymentIzipay()
    configPayment.appearance?.visualSettings?.presentationMode = .fullscreen
    configPayment.appearance?.theme = "purple"
    configPayment.appearance?.customTheme = CustomThemePaymentIzipay()
    configPayment.appearance?.logo = data["logoUrl"]

    let sensoryBrandingVisa = SensoryBranding()
    sensoryBrandingVisa.backdropColor = .white
    sensoryBrandingVisa.isSoundEnabled = true
    sensoryBrandingVisa.isHapticFeedbackEnabled = true
    sensoryBrandingVisa.checkmarkMode = .checkmark

    let sensoryBrandMastercard = MCSonicView()
    sensoryBrandMastercard.background = MCSonicBackground.white

    configPayment.appearance?.sensoryBrand = AppearenceSensoryBrandIzipay()
    configPayment.appearance?.sensoryBrand?.visaSBView = sensoryBrandingVisa
    configPayment.appearance?.sensoryBrand?.mastercardSBView = sensoryBrandMastercard

    configPayment.urlIPN = data["webhookUrl"]

    let s = UIStoryboard(name: "IziPayment", bundle: Bundle(for: PaymentFormViewController.self))
    let vc =
      s.instantiateViewController(withIdentifier: "PaymentFormView") as! PaymentFormViewController
    vc.delegate = self
    vc.configPayment = configPayment
    vc.modalPresentationStyle = .fullScreen

    guard
      let rootViewController = UIApplication.shared.connectedScenes
        .compactMap({ $0 as? UIWindowScene })
        .flatMap({ $0.windows })
        .first(where: { $0.isKeyWindow })?.rootViewController
    else {
      print("No se pudo obtener el rootViewController")
      return
    }

    DispatchQueue.main.async {
      rootViewController.present(vc, animated: true, completion: nil)
    }
  }

  public func getPaymentResult(_ paymentResult: PaymentResult) {
    let resultCode = paymentResult.code ?? ""

    if resultCode == "00" {
      let dataToSend: [String: Any?] = [
        "success": true,
        "code": paymentResult.code,
        "cardToken": paymentResult.response?.token?.cardToken,
        "cardPan": paymentResult.response?.card?.pan,
        "cardBrand": paymentResult.response?.card?.brand,
      ]
      self.sendResultToFlutter(result: dataToSend)
    } 
    else {
      let dataToSend: [String: Any?] = [
        "success": false,
        "code": paymentResult.code,
      ]
      self.sendResultToFlutter(result: dataToSend)
    }
  }

  public func executeProfiling(_ result: IzipayPayButtonSDK.ScoringParams) {

  }

  public func executeSensoryBrandVisa(view: UIView) {
    if let visaview = view as? SensoryBranding {
      DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
        visaview.animate()
      }
    }
  }

  public func executeSensoryBrandMastercard(
    view: UIView, params: IzipayPayButtonSDK.MastercardSonicParams
  ) {
    let merchant = MCSonicMerchant(
      merchantName: params.name, countryCode: params.country,
      city: nil, merchantCategoryCodes: [params.mcc], merchantId: nil)
    if let mastercardview = view as? MCSonicView, let safeMerchant = merchant {
      self.sonicManager.prepare(
        with: .soundAndAnimation, sonicCue: params.sonicCue,
        sonicEnvironment: .production, merchant: safeMerchant
      ) { status in
        if status == .success {
          self.sonicManager.play(with: mastercardview) { MCSonicStatus in
            print(MCSonicStatus)
          }
        }
      }
    }
  }
}
extension FlutterIzipayPlugin: FlutterStreamHandler {
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    self.eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    self.eventSink = nil
    return nil
  }

  public func sendResultToFlutter(result: [String: Any?]) {
    if let sink = eventSink {
      sink(result)
    }
  }
}

File: android/src/main/kotlin/com/briansalvattore/flutter_izipay/FlutterIzipayPlugin.kt
package com.briansalvattore.flutter_izipay

import android.app.Activity
import android.content.Intent
import android.util.Log
import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

import com.izipay.izipay_pw_sdk.data.model.ConfigRequest
import com.izipay.izipay_pw_sdk.data.model.OrderPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.PayOption
import com.izipay.izipay_pw_sdk.data.model.TokenPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.BillingPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.ShippingPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.AppearencePaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.AppearenceControlsPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.AppearenceVisualSettingsPaymentIzipay
import com.izipay.izipay_pw_sdk.data.model.CustomThemePaymentIzipay
import com.izipay.izipay_pw_sdk.domain.model.PaymentResponse
import com.izipay.izipay_pw_sdk.ui.fullscreend.ContainerActivity

class FlutterIzipayPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler, ActivityAware {
  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private var eventSink: EventChannel.EventSink? = null

  private var activity: Activity? = null

  companion object {
    const val REQUEST_CODE_IZIPAY = 1001

    const val IZIPAY_PAY = "pay"
    const val IZIPAY_REGISTER = "register"

    const val TAG = "FlutterIzipayPlugin"
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
    binding.addActivityResultListener { requestCode, resultCode, data ->
      handleActivityResult(requestCode, resultCode, data)
      true
    }
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel = MethodChannel(binding.binaryMessenger, "flutter_izipay/method_channel")
    methodChannel.setMethodCallHandler(this)

    eventChannel = EventChannel(binding.binaryMessenger, "flutter_izipay/event_channel")
    eventChannel.setStreamHandler(this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
        "openFormToSaveCard" -> {
            val data = call.arguments as Map<String, String>
            openFormToSaveCard(data)
        }
        "payWithCard" -> {
            val data = call.arguments as Map<String, String>
            payWithCard(data)
        }
        "payWithYape" -> {
            val data = call.arguments as Map<String, String>
            payWithYape(data)
        }
        else -> {
            result.notImplemented()
        }
    }
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
      eventSink = null
  }

  private fun sendEvent(data: Map<String, Any?>) {
      eventSink?.success(data)
  }

  private fun payWithYape(map: Map<String, String>) {
      val request = ConfigRequest(
        map["environment"] as String,
        IZIPAY_PAY,
        map["publicKey"] as String,
        map["transactionId"] as String,
        map["merchantCode"] as String,
        "",
        OrderPaymentIzipay(
          map["orderNumber"] as String, //siempre debe tener 13 dígitos
          "PEN",
          map["amount"] as String,  // siempre debe tener el formato verificar 00.00
          arrayListOf(PayOption.YAPE),
          "mobile",
          "autorize",
          map["userId"] as String, // siempre debe tener una letra
          System.currentTimeMillis().toString(),
        ),
        TokenPaymentIzipay(
          "",
        ),
        BillingPaymentIzipay(
          map["firstName"] as String,
          map["lastName"] as String,
          map["email"] as String,
          map["phoneNumber"] as String,
          map["street"] as String,
          map["city"] as String,
          map["state"] as String,
          map["country"] as String,
          map["postalCode"] as String,
          map["documentType"] as String,
          map["documentNumber"] as String,
        ),
        null,
        AppearencePaymentIzipay(
          "ESP",
          AppearenceControlsPaymentIzipay(
            true,
            false,
          ),
          AppearenceVisualSettingsPaymentIzipay(
            true,
          ),
          "purple",
          CustomThemePaymentIzipay(
            "#333399",
            "#333399",
            "#333399",
          ),
          map["logoUrl"] as String,
        ),
        map["webhookUrl"] as String,
      )

      Log.d("FlutterTapAndPayPlugin", "ConfigRequest ${request}")

      val intent = Intent(activity!!, ContainerActivity::class.java)
      intent.putExtra(ContainerActivity.REQUEST, request)
      activity!!.startActivityForResult(intent, REQUEST_CODE_IZIPAY)
  }

  private fun payWithCard(map: Map<String, String>) {
      val request = ConfigRequest(
        map["environment"] as String,
        IZIPAY_PAY,
        map["publicKey"] as String,
        map["transactionId"] as String,
        map["merchantCode"] as String,
        "",
        OrderPaymentIzipay(
          map["orderNumber"] as String, //siempre debe tener 13 dígitos
          "PEN",
          map["amount"] as String,  // siempre debe tener el formato verificar 00.00
          arrayListOf(PayOption.CARD),
          "mobile",
          "autorize",
          map["userId"] as String, // siempre debe tener una letra
          System.currentTimeMillis().toString(),
        ),
        TokenPaymentIzipay(
          "",
        ),
        BillingPaymentIzipay(
          map["firstName"] as String,
          map["lastName"] as String,
          map["email"] as String,
          map["phoneNumber"] as String,
          map["street"] as String,
          map["city"] as String,
          map["state"] as String,
          map["country"] as String,
          map["postalCode"] as String,
          map["documentType"] as String,
          map["documentNumber"] as String,
        ),
        null,
        AppearencePaymentIzipay(
          "ESP",
          AppearenceControlsPaymentIzipay(
            true,
            false,
          ),
          AppearenceVisualSettingsPaymentIzipay(
            true,
          ),
          "purple",
          CustomThemePaymentIzipay(
            "#333399",
            "#333399",
            "#333399",
          ),
          map["logoUrl"] as String,
        ),
        map["webhookUrl"] as String,
      )

      Log.d("FlutterTapAndPayPlugin", "ConfigRequest ${request}")

      val intent = Intent(activity!!, ContainerActivity::class.java)
      intent.putExtra(ContainerActivity.REQUEST, request)
      activity!!.startActivityForResult(intent, REQUEST_CODE_IZIPAY)
  }

  private fun openFormToSaveCard(map: Map<String, String>) {
      val request = ConfigRequest(
        map["environment"] as String,
        IZIPAY_REGISTER,
        map["publicKey"] as String,
        map["transactionId"] as String,
        map["merchantCode"] as String,
        "",
        OrderPaymentIzipay(
          map["orderNumber"] as String, //siempre debe tener 13 dígitos
          "PEN",
          "0.00",
          arrayListOf(PayOption.CARD),
          "mobile",
          "autorize",
          map["userId"] as String, // siempre debe tener una letra
          System.currentTimeMillis().toString(),
        ),
        TokenPaymentIzipay(
          "",
        ),
        BillingPaymentIzipay(
          map["firstName"] as String,
          map["lastName"] as String,
          map["email"] as String,
          map["phoneNumber"] as String,
          map["street"] as String,
          map["city"] as String,
          map["state"] as String,
          map["country"] as String,
          map["postalCode"] as String,
          map["documentType"] as String,
          map["documentNumber"] as String,
        ),
        null,
        AppearencePaymentIzipay(
          "ESP",
          AppearenceControlsPaymentIzipay(
            false,
            false,
          ),
          AppearenceVisualSettingsPaymentIzipay(
            true,
          ),
          "purple",
          CustomThemePaymentIzipay(
            "#333399",
            "#333399",
            "#333399",
          ),
          map["logoUrl"] as String,
        ),
        map["webhookUrl"] as String,
      )

      Log.d("FlutterTapAndPayPlugin", "ConfigRequest ${request}")

      val intent = Intent(activity!!, ContainerActivity::class.java)
      intent.putExtra(ContainerActivity.REQUEST, request)
      activity!!.startActivityForResult(intent, REQUEST_CODE_IZIPAY)
  }

  private fun handleActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
      if (requestCode == REQUEST_CODE_IZIPAY) {
        if (resultCode == Activity.RESULT_OK) {
            val dataPayLoad: PaymentResponse = Gson().fromJson(data?.getStringExtra(ContainerActivity.RESPONSEPAYLOAD).orEmpty(), object: TypeToken<PaymentResponse>(){}.type)
            
            Log.d("FlutterTapAndPayPlugin", "PaymentResponse ${dataPayLoad}")
            
            val eventData = mapOf(
              "success" to true,
              "code" to dataPayLoad.code,
              "cardToken" to dataPayLoad.response?.token?.cardToken,
              "cardPan" to dataPayLoad.response?.card?.pan,
              "cardBrand" to dataPayLoad.response?.card?.brand,
            )
            sendEvent(eventData)
        } 
        else {
            val eventData = mapOf(
              "success" to false,
            )
            sendEvent(eventData)
        }
      }
  }
}
